src/js/drawer.js

```diff
@@ -0,0 +1,139 @@
+import Toggle from 'o-toggle';
+
+const LISTEN_DELAY = 300;
+const INTENT_DELAY = 1000;
+
+function handleCloseEvents (scope, callback) {
+	let timeout;
+
+	const handleKeydown = (e) => {
+		if (e.keyCode === 27) {
+			callback();
+		}
+	};
+
+	const handleClick = (e) => {
+		if (!scope.contains(e.target)) {
+			console.log('callbacking');
+			callback();
+		}
+	};
+
+	const handleMouseenter = () => {
+		clearTimeout(timeout);
+	};
+
+	const handleMouseleave = () => {
+		timeout = setTimeout(callback, INTENT_DELAY);
+	};
+
+	const removeEvents = () => {
+		console.log('removing events');
+		clearTimeout(timeout);
+
+		scope.removeEventListener('mouseenter', handleMouseenter);
+		scope.removeEventListener('mouseleave', handleMouseleave);
+		document.removeEventListener('click', handleClick);
+		document.removeEventListener('touchstart', handleClick);
+		document.removeEventListener('keydown', handleKeydown);
+	};
+
+	const addEvents = () => {
+		console.log('adding events');
+		scope.addEventListener('mouseenter', handleMouseenter);
+		scope.addEventListener('mouseleave', handleMouseleave);
+		document.addEventListener('click', handleClick);
+		document.addEventListener('touchstart', handleClick);
+		document.addEventListener('keydown', handleKeydown);
+	};
+
+	return { addEvents, removeEvents };
+}
+
+function addDrawerToggles (headerEl, drawerEl) {
+	const controls = [];
+	controls.push(...headerEl.querySelectorAll(`[aria-controls="${drawerEl.id}"]`));
+	controls.push(...drawerEl.querySelectorAll(`[aria-controls="${drawerEl.id}"]`));
+
+	let handleClose;
+	function toggleCallback(state) {
+		if (state === 'close') {
+			handleClose.removeEvents();
+		} else {
+			// don't capture the initial click or accidental double taps etc.
+			// we could use transitionend but scoping is tricky and it needs prefixing and...
+			setTimeout(handleClose.addEvents, LISTEN_DELAY);
+		}
+
+		drawerEl.classList.toggle('o-header__drawer--closing', state === 'close');
+		drawerEl.classList.toggle('o-header__drawer--opening', state === 'open');
+	}
+
+	controls.forEach((control) => {
+		const drawerToggle = new Toggle(control, {
+			target: drawerEl,
+			callback: toggleCallback
+		});
+
+		// Both toggles have the same target, so the toggle function will be the same
+		// If there's a separate handleClose instance for each toggle, removeEvents doesn't work
+		// when the close toggle is clicked
+		if (!handleClose) {
+			handleClose = handleCloseEvents(drawerEl, drawerToggle.toggle);
```

Comment:
Ah! Yeah like you say it's a tiny bit hacky but I'm OK with it. The comment above is good 👍